jbi/services/jira.py

```diff
@@ -30,15 +31,45 @@
 instrumented_method = instrument(prefix="jira", exceptions=(errors.ApiError,))
 
 
+def log_http_error(func):
+    """A decorator to catch and log HTTP errors responses of the Jira client.
+
+    Without this the actual requests and responses are not exposed when an error
+    occurs, which makes troubleshooting tedious.
+    """
+
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        try:
+            return func(*args, **kwargs)
+        except requests.HTTPError as exc:
+            request = exc.request
+            response = exc.response
+            logger.error(
+                "HTTP: %s %s -> %s %s",
+                request.method,
+                request.path_url,
+                response.status_code,
+                response.reason,
+                extra={"body": response.text},
+            )
+            raise
+
+    return wrapper
+
+
 class JiraClient(Jira):
     """Adapted Atlassian Jira client that wraps methods in our instrumentation
     decorator.
     """
 
-    update_issue_field = instrumented_method(Jira.update_issue_field)
-    set_issue_status = instrumented_method(Jira.set_issue_status)
-    issue_add_comment = instrumented_method(Jira.issue_add_comment)
-    create_issue = instrumented_method(Jira.create_issue)
+    update_issue_field = log_http_error(instrumented_method(Jira.update_issue_field))
+    set_issue_status = log_http_error(instrumented_method(Jira.set_issue_status))
+    issue_add_comment = log_http_error(instrumented_method(Jira.issue_add_comment))
+    create_issue = log_http_error(instrumented_method(Jira.create_issue))
+    get_project_components = log_http_error(
+        instrumented_method(Jira.get_project_components)
+    )
```

Comment:
Could we get away with only monkeypatching the base class's [`raise_for_status`](https://github.com/atlassian-api/atlassian-python-api/blob/32b5af22ad3702a046b6184506fc1b5bd98a7b40/atlassian/rest_client.py#L462) method?

```suggestion
    raise_for_status = log_http_error(Jira.raise_for_status)
    update_issue_field = instrumented_method(Jira.update_issue_field)
    set_issue_status = instrumented_method(Jira.set_issue_status)
    issue_add_comment = instrumented_method(Jira.issue_add_comment)
    create_issue = instrumented_method(Jira.create_issue)
    get_project_components = instrumented_method(Jira.get_project_components)
```

Alternatively, we could write our own `raise_for_status` since we'd potentially double log the error in [this case](https://github.com/atlassian-api/atlassian-python-api/blob/master/atlassian/rest_client.py#LL479C1-L481C44):

```
except Exception as e:
    log.error(e)
    response.raise_for_status()
```

Reply:
Yes, you're right, let's decorate `raise_for_status()`, it's way more elegant!

> since we'd potentially double log the error in [this case](https://github.com/atlassian-api/atlassian-python-api/blob/master/atlassian/rest_client.py#LL479C1-L481C44):

Yes, this `Exception` is raised when the response cannot be parsed into JSON or when the parsed content does not have the right format. If I'm not wrong, we would indeed have two calls to `logger.error()`, but not with the same error. One with the JSON/parsing stuff, and one with the HTTP status.